### ISSUE TITLE
[Feature Request] Detect unbeartyped @dataclasses and do something about it

### ISSUE THREAD

**Issue by denisrosset** [beartype/beartype#119]
My understanding is that dataclass support works by decorating the `__init__` method. However, dataclasses are not checked when they are invalid and passed as arguments to other functions. Nor when `beartype.abby.die_if_unbearable` is called.

I wrote a small test suite, shown below.

What would be needed to make this support happen?

```python
from dataclasses import dataclass

import beartype
import beartype.abby
import beartype.roar
import beartype.vale
import pytest
from typing_extensions import Annotated, get_type_hints

@beartype.beartype
@dataclass(frozen=True)
class WithDecorator:
    a: Annotated[int, beartype.vale.Is[lambda x: x >= 0]]


@dataclass(frozen=True)
class WithoutDecorator:
    a: Annotated[int, beartype.vale.Is[lambda x: x >= 0]]


def test_beartype_decorator() -> None:
    with pytest.raises(beartype.roar.BeartypeException):
        WithDecorator(-1)


def test_unbearable_dataclass() -> None:  # fails
    with pytest.raises(beartype.roar.BeartypeException):
        data = WithoutDecorator(-1)
        beartype.abby.die_if_unbearable(data, WithoutDecorator)


def test_dataclass_in_argument() -> None:  # fails
    with pytest.raises(beartype.roar.BeartypeException):

        @beartype.beartype
        def fun(d: WithoutDecorator) -> None:
            pass

        data = WithoutDecorator(-1)
        fun(data)


def test_unbearable_field() -> None:
    hints = get_type_hints(WithoutDecorator, include_extras=True)
    with pytest.raises(beartype.roar.BeartypeException):
        data = WithoutDecorator(-1)
        beartype.abby.die_if_unbearable(data.a, hints["a"])
```

**Comment by leycec** [comment]
*W00t, Waterloo!* `</ahem>`

> My understanding is that dataclass support works by decorating the `__init__` method.

[I see your understanding is as large as mine.](https://www.youtube.com/watch?v=pPkWZdluoUg)

> However, dataclasses are not checked when they are invalid and passed as arguments to other functions. Nor when `beartype.abby.die_if_unbearable` is called.

Indeed, the ugly truth is now exposed for all too see. Like `@beartype`, the `@dataclass.dataclass` decorator is really just an obscene pile of runtime trickery. It's unclear whether @beartype can (or even should) automatically insinuate itself into that trickery by pretending third-party dataclasses not originally decorated by `@beartype` were decorated by `@beartype` after all.

Thankfully, this is Python. We can do whatever we like regardless of what anyone else thinks, because that's what the world's second-slowest language buys us. Options present themselves like ripe sushi for the plucking:

1. **Just do it manually.** Python decorators are actually just standard callables, which means you can call them just as you would any other function. Pretty sure this works, so let's pretend I actually tested this:

```python
def test_unbearable_dataclass() -> None:  # fails
    with pytest.raises(beartype.roar.BeartypeException):
        data = WithoutDecorator(-1)
        WithDecoratorAfterTheFact = beartype.beartype(WithoutDecorator)
        beartype.abby.die_if_unbearable(data, WithDecoratorAfterTheFact)
```

I couldn't help myself and tested an equivalent snippet from within an IPython REPL. I proudly confirm worky:

```python
>>> from dataclasses import dataclass
>>> import beartype
>>> import beartype.abby
>>> import beartype.vale
>>> from typing_extensions import Annotated
>>> @dataclass(frozen=True)
... class WithoutDecorator:
...     a: Annotated[int, beartype.vale.Is[lambda x: x >= 0]]
>>> WithDecoratorAfterTheFact = beartype.beartype(WithoutDecorator)
>>> data = WithoutDecorator(-1)
>>> beartype.abby.die_if_unbearable(data, WithDecoratorAfterTheFact)
beartype.roar.BeartypeCallHintParamViolation: @beartyped WithoutDecorator.__init__() parameter a=-1 violates type hint typing.Annotated[int, Is[lambda x: x >= 0]], as -1 violates validator Is[lambda x: x >= 0]:
    False == Is[lambda x: x >= 0].
```

![roar, bro](https://www.lamontagneart.com/wp-content/uploads/2019/09/Roar.jpg)

2. **Shift the burden of proof onto @beartype.** Arguably, @beartype itself should either:
* Emit a **non-fatal warning** when passed a dataclass *not* decorated by @beartype. That could get kinda annoying, though. So...
* Silently decorate dataclasses *not* decorated by @beartype with @beartype. That gets non-trivial fast, though. So, let's dig deeper into this special madness.

## Automate It, Bro

Silently decorating dataclasses *not* decorated by @beartype with @beartype reduces annoyance, but comes at the cost of **probably dramatically harming time performance if implemented naïvely.** Why? Because the naïve implementation would dynamically declare one new @beartype-decorated dataclass for each @beartype call passed an undecorated dataclass. Obviously, that's awful.

If we go down this road, we should go down this road crazy fast without regard for local traffic laws – otherwise, what's the point, right? This. Is. @beartype.

So, the non-naïve implementation requires @beartype to internally cache and reuse @beartype-decorated dataclasses when passed undecorated dataclasses. Obviously, that trades off time for speed – because @beartype would now be effectively doubling the number of dataclasses declared per Python process.

Maybe nobody cares? I have no data points here. If somebody cares, LRU caching shyly raises its hand at the back of the class. And then there's always...

3. **Shift the burden of proof onto @beartype but in a crazy way.** Hacky examples include your concluding `beartype.abby.die_if_unbearable(data.a, hints["a"])` test, in which @beartype would iteratively check all dataclass fields of undecorated dataclasses. That probably sounds great on paper, except that:
  * It requires substantially more heavy lifting from @beartype – heavy lifting we don't necessarily have time to implement. [Time, what is time!?!?](https://www.youtube.com/watch?v=M5KXDzi1iVw)
  * It possibly imposes runtime inefficiencies I shudder to even fathom in my darkest dreams.

## tl;dr

Option \#1. for short-term gain. Option \#2 to minimize long-term pain – hopefully with caching. High fives all around! :hand:

**Comment by dertilo** [comment]
# Hi! I am using the following hack to beartype all "my" dataclasses' `__init__` -methods
* how bad/dangerous/suboptimal is this hack?
* I guess my motivation to "hack" the "dataclass"-decorator (over a year ago?) was somehow related to issues with my IDE (pycharm) that was/is unwilling to "handle" the `@beartype` ontop of `@dataclass`

### 1. hack-patch python3.9/dataclasses.py
* should be more or less the same for other python versions (?)
* the one-line `wrap`-function gets "expanded" to beartyp the dataclasses init-method if it `is_one_of_my_own`
* to calculate the `is_one_of_my_own`-bool might be better to use some trie/caching?
```diff
--- original.py 2023-03-17 07:17:55.333675520 +0100
+++ patched.py  2023-03-17 07:17:56.989660276 +0100
@@ -1,4 +1,4 @@
-def dataclass(
+def beartyped_dataclass(
     cls=None,
     /,
     *,
@@ -22,7 +22,17 @@
     """
 
     def wrap(cls):
-        return _process_class(cls, init, repr, eq, order, unsafe_hash, frozen)
+        data_cls = dataclasses._process_class(
+            cls, init, repr, eq, order, unsafe_hash, frozen
+        )
+        this_dataclasses_file = sys.modules[cls.__module__].__file__
+        is_one_of_my_own = any(
+            (this_dataclasses_file.startswith(s) for s in BEARTYPED_DATACLASS_PREFIXES)
+        )
+        if is_one_of_my_own:
+            data_cls.__init__ = beartype(data_cls.__init__)  # type: ignore[assignment]
+        return data_cls
 
     # See if we're being called as @dataclass or @dataclass().
     if cls is None:

dataclasses.dataclass = beartyped_dataclass
```

### 2. define which dataclasses to beartype
* `BEARTYPED_DATACLASS_PREFIXES`: a collection of path/directory prefixes that defines which dataclasses are to be beartyped
 
```python
BEARTYPED_DATACLASS_PREFIXES: set[str] = set()

def beartype_all_dataclasses_of_this_files_parent(file: str):
    package_dir = str(pathlib.Path(file).parent.resolve())

    already_contained = any(
        (package_dir.startswith(s) for s in BEARTYPED_DATACLASS_PREFIXES)
    )
    if not already_contained:
        BEARTYPED_DATACLASS_PREFIXES.add(package_dir)

    # maybe remove children/prefixes (subdirectories)
    children = [
        p
        for p in BEARTYPED_DATACLASS_PREFIXES
        if p.startswith(package_dir) and p != package_dir
    ]
    is_parent = len(children) > 0
    if is_parent:
        for ch in children:
            BEARTYPED_DATACLASS_PREFIXES.remove(ch)
        BEARTYPED_DATACLASS_PREFIXES.add(package_dir)


```
* to beartype all dataclasses of a package one needs to add 2 lines to the `__init__.py`
```python
from misc_utils.beartyped_dataclass_patch import beartype_all_dataclasses_of_this_files_parent

beartype_all_dataclasses_of_this_files_parent(__file__)
```
I am super grateful for any feedback!

**Comment by leycec** [comment]
@dertilo!! Everybody, it's our newest favourite GitHub Sponsor!! @beartype :heart_eyes_cat: you so much.

`</ahem>`

...where was I? Where am I? Who am I? Oh. Right. GitHub. Let's do this on a Saturday night. 

> my IDE (pycharm) that was/is unwilling to "handle" the `@beartype` ontop of `@dataclass`

Curse you a hundred-fold, JetBrains. Curse you.

My wife and I otherwise *love* PyCharm, but JetBrains made the curious decision to **implement their own ad-hoc Python static type-checker inside PyCharm.** Every other IDE I've used just reuses standard off-the-shelf static type-checkers, because that is the easy (and therefore profitable) thing to do.

For example, let us consider the most popular Python IDE. I speak of Microsoft's VSCode, which supports Python via Pylance, which:

* Internally reuses Microsoft's own open-source `pyright` static type-checker.
* Optionally supports mypy as well.

@beartype officially supports both `pyright` and mypy and thus just works "out-of-the-box" with VSCode. Personally, I prefer command-line Vim, because I live in a cabin in the woods. People like me like insanity like Vim.

Sadly, it's probably impossible for @beartype to support PyCharm's non-standard (and probably closed-source) static type-checker. It hurts! We really *want* to support that static type-checker – whatever it is. It doesn't even seem to have a name, which is equally bizarre. Sad cat is cat. :crying_cat_face: 

> how bad/dangerous/suboptimal is this hack?

You are monkey-patching the Python standard library in compelling and explosive ways. The answer, of course, is... *this is so awesome!* Here at @beartype Studios, we strongly support ~~risky, reckless, and lawless~~ bold, brave, and innovative behaviour like this.

You're in luck, too. Line-by-line inspection of your helpful diff suggests that you've done everything absolutely right! Well, *almost* everything. You violate privacy encapsulation by directly calling `_process_class()`, which isn't great. Although Python 3.11 still has that private function, it now accepts *way* more parameters.

Here's what the same `dataclasses` logic looks like under Python 3.11:

```python
# Here's the Python 3.11 variant of `wrap()`:
    def wrap(cls):
        return _process_class(cls, init, repr, eq, order, unsafe_hash,
                              frozen, match_args, kw_only, slots,
                              weakref_slot)

# Here's the Python 3.9 variant of `wrap()`:
     def wrap(cls):
         return _process_class(cls, init, repr, eq, order, unsafe_hash, frozen)
```

*Super*-different, right? This means that your current approach, while extraordinarily clever, is also non-portable. Which isn't great.

An alternative approach would be to black-box this *without* violating privacy encapsulation. Consider something resembling:

```python
def beartype_dataclass(*args, **kwargs):
    return beartype(dataclass(*args, **kwargs))

dataclasses.dataclass = beartyped_dataclass
```

:laughing: :rofl: 

Ridiculous, right!?!? There's *no* way that possibly works. But... it does! It actually works! Rejoice, for Spring is truly upon us:

```pycon
>>> from dataclasses import dataclass
>>> @dataclass
... class SpringIsEverywhere(object):
...     in_every_sight: str
...     and_every_sound: int
>>> rejoice_everybody = SpringIsEverywhere(
...     in_every_sight='This is a string.',
...     and_every_sound='Pretty sure this is not an integer.',
... )
beartype.roar.BeartypeCallHintParamViolation: @beartyped
__main__.SpringIsEverywhere.__init__() parameter and_every_sound='Pretty sure
this is not an integer.' violates type hint <class 'int'>, as str 'Pretty sure
this is not an integer.' not instance of int.
```

As we see, @beartype correctly accepts the first valid parameter but rejects the second invalid parameter. This is shocking. This is @beartype.

Everything else looks sane. Well, `beartype_all_dataclasses_of_this_files_parent()` is pretty intense; you've *sorta* implemented a crude alternative to import hooks. That's not necessarily bad, of course. Your approach clearly works. Clearly working is really all that matters.

Tremendous work, bro! And you'll be delighted to learn that @beartype will [implement import hooks in some future version to be released by 2057 (i.e., the year the Big Asteroid finally hits Earth)](https://github.com/beartype/beartype/issues/43). Once released, you won't need to do *any* of the above monkey-patching anymore. Indeed, you won't need to reference `@beartype` anywhere at all in your codebase anymore except for a single statement at the top of your top-level `{your_package}.__init__` module resembling:

```python
# Import the requisite machinery
from beartype.claw import beartype_all

# Automatically @beartype everything, everywhere, for all time.
beartype_all()
```

I've even already implemented like 95% of the `beartype.claw` subpackage. Then I got tired, went to bed, and stumbled off to do other stuff. I never did finish `beartype.claw`... but I will someday! I swear this, @dertilo. And Berlin will never be the same again.
